learning/modules/computer-science/shors-algorithm.ipynb

```diff
@@ -0,0 +1,1202 @@
+    "M_2^2|1\\rangle &= |4\\rangle \\\\\n",
+    "M_2^3|1\\rangle &= |8\\rangle \\\\\n",
+    "M_2^4|1\\rangle &= |1\\rangle \\\\\n",
+    "\\end{aligned}\n",
+    "$$\n",
+    "\n",
+    "So, the order $r=4$. The eigenstates we're interested in will be an equal superposition of all of the states that were cycled through above, with various phases:\n",
+    "\n",
+    "$$\n",
+    "\\begin{aligned}\n",
+    "|\\psi_0\\rangle &= \\frac{1}{2}(|1\\rangle+|2\\rangle+|4\\rangle+|8\\rangle) \\\\\n",
+    "|\\psi_1\\rangle &= \\frac{1}{2}(e^{2 \\pi i \\frac{0}{4}}|1\\rangle+e^{2 \\pi i \\frac{1}{4}}|2\\rangle+e^{2 \\pi i \\frac{2}{4}}|4\\rangle+e^{2 \\pi i \\frac{3}{4}}|8\\rangle) \\\\\n",
+    "&= \\frac{1}{2}(|1\\rangle+i|2\\rangle-|4\\rangle-i|8\\rangle) \\\\\n",
+    "|\\psi_2\\rangle &= \\frac{1}{2}(e^{2 \\pi i \\frac{0}{4}}|1\\rangle+e^{2 \\pi i \\frac{2}{4}}|2\\rangle+e^{2 \\pi i \\frac{4}{4}}|4\\rangle+e^{2 \\pi i \\frac{6}{4}}|8\\rangle) \\\\\n",
+    "&= \\frac{1}{2}(|1\\rangle-|2\\rangle+|4\\rangle-|8\\rangle) \\\\\n",
+    "|\\psi_3\\rangle &= \\frac{1}{2}(e^{2 \\pi i \\frac{0}{4}}|1\\rangle+e^{2 \\pi i \\frac{3}{4}}|2\\rangle+e^{2 \\pi i \\frac{6}{4}}|4\\rangle+e^{2 \\pi i \\frac{9}{4}}|8\\rangle) \\\\\n",
+    "&= \\frac{1}{2}(|1\\rangle-i|2\\rangle-|4\\rangle+i|8\\rangle) \\\\\n",
+    "\\end{aligned}\n",
+    "$$\n",
+    "\n",
+    "</details>\n",
+    "\n",
+    "So, say we were able to initialize our qubit state into one of these eigenstates (spoiler — we're not. Or, at least not easily. We'll explain why and what we can do instead shortly). Then we could use QPE to estimate the corresponding eigenvalue, $\\omega_j = e^{2 \\pi i \\theta_j}$ where $\\theta_j = \\frac{j}{r}$. Then, we'll be able to determine the order $r$ by the simple equation:\n",
+    "\n",
+    "$$r = \\frac{j}{\\theta_j}.$$\n",
+    "\n",
+    "But remember, I said that QPE *estimates* $\\theta_j$ — it doesn't give us an exact value. We need the estimate to be good enough to differentiate between $r$ and $r+1$. The more control qubits $m$, we have, the better the estimate will be. In the problems at the end of the lesson, you'll be asked to determine the minimum $m$ needed to factor a number $N$.\n",
+    "\n",
+    "Now, we have to reconcile a problem. All of the above explanation of how to find $r$ begins with preparing the eigenstate $|\\psi_j\\rangle = \\tfrac{1}{\\sqrt{r}}\\sum_{k=0}^{r-1}{e^{\\frac{2 \\pi i j k}{r}} |a^k \\rangle}$. But we don't know how to do that without already knowing what $r$ is. The logic is circular. We need a way to estimate the eigenvalue *without* initializing the eigenstate.\n",
+    "\n",
+    "Instead of starting with an eigenstate of $M_a$, we can prepare the initial state into the $n$-qubit state corresponding to $|1\\rangle$ in binary (i.e. $|000...01\\rangle$). Although this state itself is obviously not an eigenstate of $M_a$, it is a superposition over all of the eigenstates $|\\psi_k\\rangle$:\n",
```

Comment:
```suggestion
    "Instead of starting with an eigenstate of $M_a$, we can prepare the initial state into the $n$-qubit state corresponding to $|1\\rangle$ in binary (as in, $|000...01\\rangle$). Although this state itself is obviously not an eigenstate of $M_a$, it is a superposition over all of the eigenstates $|\\psi_k\\rangle$:\n",
```